chore: Separate lt and ht pickup and dropoff

sfoster1 opened the pull request:
This PR looks gigantic but most of it is little fiddly fixups to pipette config files that are schema checked and don't change data.

Our 96 channel pipettes pick up full racks of tips using a cam grabber, but they're also capable of picking up less tips than that through the normal press-fit routine that the low throughput pipettes use. Those routines need different configuration values that do not overlap, and we previously had them overlapping.

This commit
- Changes the pipette config general schema (this is a good place to start when reviewing) to fully separate the two styles of pickup (press fit and cam action) and the two styles of drop tip (plunger eject and cam action). Low throughput pipettes do press fit and plunger eject; high throughput pipettes do both kinds of pickup, and cam action drop. Current configuration is folded in to the two new locations.
- Changes all the other code to adapt to that
- Changes the pipette geometry schema to add the rows and columns in which the nozzles are arranged so we can do partial tip pickup calculations based on it
- Changes the partial tip pickup stuff to work based on symbolic geometry instead of points

## Review requests
- Check that no data changed except in https://github.com/Opentrons/opentrons/commit/a9ec6395e4fa259180d40863c56514625aab37da
- Do the new configs make a little more sense

## Testing
- [x] Pick up some tips!

sfoster1 committed:
chore(shared-data): pipette: refactor pickup, drop

These configs had a couple things to change:
- pickup and drop actually represent configuration for two different
actions: the press fit / plunger eject sequence that the low throughput
pipettes do, and the cam action sequence that the high throughput
pipettes do when handling a full rack. information was actually
overloaded and lost: there was only room for one speed for both separate
actions. By separating the configs for each action, we get that
information back
- fold in the per tip current information into the pickup config so it
looks a bit more natural and also can be distributed between the action kinds

sfoster1 committed:
chore(api): adapt to new tip configs

Now that we have separate configuration elements for picking up tips
with press fit vs cam action and dropping them with plunger eject and
cam action, we need to use them in a way that is nicely integrated with
the nozzle configuration stuff. That means that the plan methods should
check the current nozzle configuration and use the appropriate config
and behavior.

sfoster1 committed:
chore(robot-server): use partial tip pickup in cal

When we're calibrating an OT-2 multi channel, we do it with just one
tip. This means we have to drop the current to have the tip attached
with the appropriate amount of force. We now have a bunch of other code
focused on doing exactly that tied up with the partial tip pickup
mechanics, so use it instead.

sfoster1 committed:
chore(shared-data): 96: correcter partial currents

These are good for 1, 8, and 96 but probably not anything else.

sfoster1 committed:
3am going back for more schema refactor

sfoster1 committed:
update hardware-testing

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -150,31 +150,89 @@ class PlungerHomingConfigurations(BaseModel):
-class TipHandlingConfigurations(BaseModel):
+class PressFitPickUpTipConfiguration(BaseModel):
     presses: int = Field(
-        default=0.0, description="The number of tries required to force pick up a tip."
```
Can we keep the previous descriptions for these fields? They make more sense imo.

sfoster1 replied:
That's not what that is, though. You're going incrementally deeper every time, it's not just a retry

Laura-Danielle replied:
Added suggestions below.

sanni-t commented on shared-data/pipette/schemas/2/pipettePropertiesSchema.json:
```diff
@@ -162,42 +158,124 @@
-        "increment": { "$ref": "#/definitions/positiveNumber" },
-        "distance": { "$ref": "#/definitions/positiveNumber" }
-      }
+      "anyOf": [
```
Why `anyof` and not `oneof`?

sfoster1 replied:
Because you're allowed to have both! The 96, for instance, has both: it needs to be able to do press-fit tip pickup for partial tip configurations, and needs to be able to do cam-action tip pickup for full configurations.

sanni-t replied:
Yea, I totally had a brain fart moment

Laura-Danielle reviewed (approved):
Code in general looks good, just had a comment about some updated field descriptions.

I think also we'll want to do some extra testing on mutable configurations since there are a lot of changes there.

sanni-t commented on shared-data/pipette/definitions/2/general/ninety_six_channel/p1000/3_5.json:
```diff
@@ -44,22 +70,7 @@
-      "48": 0.75,
-      "96": 1.5
-    }
+    "availableConfigurations": [1, 8, 12, 16, 24, 48, 96]
```
Interesting, why are the other versions limited to [1, 8, 12, 96] configurations while this one has all available? Is there something in the old versions that limits their use of other partial configurations?

Laura-Danielle replied:
Human error 😆

sfoster1 replied:
@Laura-Danielle Human error like "3.5 should also only have [1, 8, 12, 96] available" or human error like "all versions should have 24 and 48"?

Laura-Danielle replied:
Human error as in I forgot about 16, 24, 48 as tip options. To be fair though, we don't really use this config as much as I thought we would.

sfoster1 replied:
will add them

ahiuchingau reviewed (approved):
THIS IS FANTASTIC! THANK YOU!!

What does cam stand for?

sanni-t reviewed (approved):
Changes make sense to me. Just a couple of questions.
Sounds like this will need thorough testing of all pipettes? (even OT2) Let me know if you or @caila-marashaj need any help with that. The 96 channel partial tip protocols we've tested so far are [here](https://github.com/Opentrons/test-protocols/tree/main/abr_testing/96_channel_partial_tip_protocols)

sfoster1 committed:
fix(api,shared-data): Use symbols for partial tips

We had implemented partial tip configurations by doing math on the
offsets of all the nozzles. This works, but has two downsides:
1. It's sort of hard to read unless you have a good mental model of the
physical layouts
2. It relies on doing exact numerical comparisons

In general, I think it is never a good idea to try and go _from_ numbers
_to_ symbolic knowledge. It's too vulnerable to small perturbations, and
it's too annoying to read.

Instead, let's encode the structure of the pipette nozzles in the
pipette geometry defs by writing down the rows and columns of the
pipette layout. Then, we can do our partial configuration logic based on
the rows and columns of the pipette.

We now do not hit a bug when we request a partial configuration that
doesn't include A1; and we now do not rely on array stride math; and we
now technically support arbitrary axis-aligned rectangles (at least on
this layer - definitely verboten above).

sfoster1 committed:
js lint

sfoster1 committed:
fix some python lint issues

Laura-Danielle commented on api/src/opentrons/hardware_control/nozzle_manager.py:
```diff
@@ -26,60 +48,81 @@ class NozzleConfigurationType(Enum):
+                if len(current_cols) == len(physical_cols) // 2:
+                    return NozzleConfigurationType.QUADRANT
+
+        return NozzleConfigurationType.SUBRECT
```
Can we just remove `QUADRANT` in favor of `SUBRECT` since a quadrant is a a subrectangle anyways? I can't see a need to distinguish between the two.

sfoster1 replied:
We allow actual quadrants but not arbitrary subrects though, right? I feel like it's worth being able to distinguish for that reason

Laura-Danielle replied:
No I've been referring to quadrant as representing arbitrary sub-rectangles. We're never actually verifying (until here I guess) that it is truly a quadrant, but I don't think we really _need_ to do that. I'm currently working on boundary box checks again -- finally -- and I can't see any need to distinguish between the two.

sfoster1 replied:
Okay!

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -359,6 +359,16 @@ class Config:
         }
 
 
+class PipetteRowDefinition(BaseModel):
```
Can we just call this `PipetteRowColumnDefinition` or something to that effect and get rid of the separate definitions for the two? They have the exact same fields.

sfoster1 replied:
They have the same fields but those fields do not have the same types in the json, and the only reason they have the same type here is because our version of pydantic isn't capable of doing fine enough divisions. I actually wrote a whole validator for this before realizing I did it based on future pydantic docs that we can't use, and as soon as we update I'll be adding them again. Specifically, the row names and column names I'd like to be separate and verified independently.

Laura-Danielle replied:
I mean they have the same basic type with some pattern validation in JSON. If you want to split them off here already, then I'd add the validation checks in these pydantic models.

sfoster1 replied:
They do not have the same pattern validation. The column names look for `/[0-9]+/` and the row names look for `/[A-Z]+/`. I'll add the validation

sfoster1 replied:
I think we should bias to the side of keeping more complex constraints in the jsonschema, though. It's just so much better at it.

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -369,6 +379,8 @@ class PipetteGeometryDefinition(BaseModel):
     )
     nozzle_map: Dict[str, List[float]] = Field(..., alias="nozzleMap")
+    ordered_columns: List[PipetteColumnDefinition] = Field(..., alias="orderedColumns")
+    ordered_rows: List[PipetteRowDefinition] = Field(..., alias="orderedRows")
```
We should probably add validation on these two fields as well similar to `nozzle_map`.

sfoster1 replied:
I want to but I want to do it inside the submodules and I want to wait until we're on a higher version of pydantic because it will be 99 times easier

sfoster1 replied:
ended up doing it

Laura-Danielle commented on api/src/opentrons/hardware_control/nozzle_manager.py:
```diff
@@ -95,13 +138,8 @@ def xy_center_offset(self) -> Point:
-        difference = self.map_store[self.back_left] - self.map_store[self.front_right]
-        y_rows_length = int(difference[1] // INTERNOZZLE_SPACING)
-        return map_store_list[starting_idx + y_rows_length]
+        front_left = next(iter(self.columns.values()))[-1]
```
This technically is only the 'front left' of the 8 channel. It would give an incorrect value if the 96 channel was in any configuration that did not include the leftmost nozzle. Should maybe just determine this in the `build` function and save.

sfoster1 replied:
So actually, and I think that maybe I should clarify this more, in an individual `NozzleMap` instance the `map_store`, `rows`, and `columns` actually only refer to the tips that are valid in the map. The first element in columns will _always_ be the left-most column of this specific partial tip configuration, therefore the last element of that first column will always be the frontest, leftest channel. This is verified in the `test_*_pipette_map_entries` tests in `test_nozzle_manager`.

Laura-Danielle replied:
yeah the frontest/leftest channel of _that_ configuration, not of the whole pipette -- whereas previously it would only ever refer to the front left-most? (H1) of the pipette.

I can't think of an exact use-case where it could be a problem to change that behavior, but something to look out for/be aware of in case there is weird behavior somewhere. I would also change the comment inside the property to be a little more explicit about what's happening.

sfoster1 replied:
Sure. I think the idea is that if you want the front-left most of the pipette you would get it from the physical nozzlemap instance that the manager keeps around.

Laura-Danielle commented on api/src/opentrons/hardware_control/nozzle_manager.py:
```diff
@@ -110,87 +148,123 @@ def tip_count(self) -> int:
+                wrapping=[PythonException(e)],
+            ) from e
+
+        def _rows_in_map() -> Iterator[Tuple[str, List[str]]]:
```
imo I think this would be cleaner if we just took the first/last index from front/back row/column and then used list splicing.

Laura-Danielle replied:
less lines of code too :D

sfoster1 replied:
Yes, but that would require (if I'm understanding right) parsing the row and column from the name of the well, which is something I tried to avoid as much as I could in favor of keeping the well names opaque atomic artifacts, which is an approach that feels more robust. For instance, if we wanted to, there would be nothing preventing us from naming a tip location just `1` or for that matter `steve` in this implementation - they're fully arbitrary. 

The iteration process also keeps a robustness element of handling arbitrarily and possibly irregular pipettes - because we're iterating linearly, we don't actually care how many things we emit, and that's another thing that isn't really necessary but that I like a lot about this.

Laura-Danielle replied:
No you shouldn't need to.

You could just return the back_left_row/back_left_column as a tuple of (name, contents) rather than just name. Then you could do:

```
list(ordered_dict.items()).index((name, contents))
```

As long as that matches the internals then you will get the correct index. I think this should also take care of irregular pipette patterns, unless I don't understand what types of irregularities you have in mind.

sfoster1 replied:
Okay. It wouldn't handle the kind of irregularity where you have rows with different numbers of columns though. Which we don't have.

sfoster1 replied:
Oh ok I get it. Yeah I can do that. It won't support irregular format pipettes, like if rows had different number of columns, but we won't have those anyway probably.

sfoster1 replied:
@Laura-Danielle how does https://github.com/Opentrons/opentrons/pull/13962/commits/39fe1515e6bbf119d520d88efc93b26f1ab3e66a look

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -150,31 +150,89 @@ class PlungerHomingConfigurations(BaseModel):
+        ..., description="The incremental amount to press more by each press"
+    )
+    distance: float = Field(
+        ..., description="The distance to press on the first stroke"
```
```suggestion
        ..., description="The starting distance to begin a pick up tip from"
```

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -150,31 +150,89 @@ class PlungerHomingConfigurations(BaseModel):
+        description="The number of times to press (incrementally more each time by increment)",
+    )
+    increment: float = Field(
+        ..., description="The incremental amount to press more by each press"
```
```suggestion
        ..., description="The increment to move the pipette down on each force tip pickup press"
```

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -150,31 +150,89 @@ class PlungerHomingConfigurations(BaseModel):
     presses: int = Field(
-        default=0.0, description="The number of tries required to force pick up a tip."
+        ...,
+        description="The number of times to press (incrementally more each time by increment)",
```
```suggestion
        description="The number of times to force pickup (incrementally more each time by increment)",
```

Laura-Danielle commented on shared-data/python/opentrons_shared_data/pipette/pipette_definition.py:
```diff
@@ -150,31 +150,89 @@ class PlungerHomingConfigurations(BaseModel):
+    distance: float = Field(
+        ..., description="The distance to press on the first stroke"
+    )
+    speed: float = Field(..., description="How fast to press each stroke")
```
```suggestion
    speed: float = Field(..., description="The speed to move the Z axis for each force pickup")
```

Laura-Danielle reviewed (commented):
Will not request changes as to ensure I don't block this PR from being merged. However, I do have a few comments that I think we should address before merging.

sfoster1 committed:
Update shared-data/python/opentrons_shared_data/pipette/pipette_definition.py

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>

sfoster1 committed:
Update shared-data/python/opentrons_shared_data/pipette/pipette_definition.py

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>

sfoster1 committed:
Update shared-data/python/opentrons_shared_data/pipette/pipette_definition.py

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>

sfoster1 committed:
Update shared-data/python/opentrons_shared_data/pipette/pipette_definition.py

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>

sfoster1 committed:
format after changes

ryanthecoder commented on shared-data/pipette/definitions/2/general/eight_channel/p1000/3_5.json:
```diff
@@ -4,14 +4,28 @@
+      "distance": 13.0,
+      "currentByTipCount": {
+        "1": 0.2,
+        "2": 0.14,
```
Is this one meant to be lower than everything else

sfoster1 replied:
Yeah, these are sort of known-to-be-messed-up first pass numbers

sfoster1 committed:
add pydantic validators for row and col names

sfoster1 committed:
update docstrings for nozzle maps

sfoster1 committed:
get rid of QUADRANT

sfoster1 committed:
use slices instead of iteration for style reasons

sfoster1 committed:
add missing partial configs

Laura-Danielle reviewed (approved):
nice, lgtm!
